### PR TITLE
Claudia/5562

### PR DESCRIPTION
**Pull request by chemaper**
#### Reviewers

 @flopezag @jframos 
#### Description
- Scripts to create, test and publish GE images from base image + create script + test script
#### Testing
- Code installed and used in 130.206.112.2 host

### CODE REVIEW

File: scripts/support/create_ge_images/README.rst [hunk 1]
```diff
@@ -0,0 +1,159 @@
+*******************************************
+Scripts to create, test & publish GE images
+*******************************************
+
+This is the software use to implement the functionality described at
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: software used

File: scripts/support/create_ge_images/README.rst [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+https://forge.fiware.org/plugins/mediawiki/wiki/testbed/index.php/FIWARE_LAB_Image_Deployement_Guideline.
+
+That is, this software creates, tests and publish an image, using the pair of
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: publishes

File: scripts/support/create_ge_images/README.rst [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+* kvm
+* nova python client
+
+The scripts should be copied to */opt/create_ge_images* to share the installation
+between several users. Also the module *osclients.py*, provided in the parent
+folder, must be copied.
+
+Each user must create the folder *~/create_ge_images* and load the environment
+variables (``OS_USERNAME``, ``OS_PASSWORD``, ``OS_TENANT_NAME``, ``OS_AUTH_URI``) with its
+FiwareLab credential. The script *sec_group* is useful to initialize the
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: FIWARE Lab

File: scripts/support/create_ge_images/README.rst [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+user account: it creates a pair of security groups (one of them to create
+the image and the other one to test it), generates and registers a key pair and allocates a
+floating IP if it is not already allocated. The scripts expects that the first
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: The scripts wait...    or    The script waits...

File: scripts/support/create_ge_images/README.rst [hunk 5]
```diff
@@ -0,0 +1,159 @@
+*******************************************
+Scripts to create, test & publish GE images
+*******************************************
+
+This is the software use to implement the functionality described at
+https://forge.fiware.org/plugins/mediawiki/wiki/testbed/index.php/FIWARE_LAB_Image_Deployement_Guideline.
+
+That is, this software creates, tests and publish an image, using the pair of
+scripts, a create script and a test script, provided by each GE owner.
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: creation script, testing script

File: scripts/support/create_ge_images/README.rst [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+as parameter the name of the image. For example, to create the image
+*myimage* using the *debian* image, the command to invoke is:
+
+.. code::
+
+  /opt/create_ge_images/create_template_debian.sh myimage
+
+Inside the folder myimage, a pair of logs will be generated: *create.log* is the
+result of invoking the *create.sh* script and *test.log* is the output of
+the *test.sh* launch. Of course, if the execution of the create script fails,
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: creation script

File: scripts/support/create_ge_images/README.rst [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+The last phase is to check the image. A server is instantiated, using again the
+generated keypair and assigning the floating IP. This time, however, the security
+group *allopen* is used and the name of the VM is the same than the image but
+with the *test-* prefix. A ssh-agent is started to insert the generated public key.
+Then the test script is invoked. If it complete without errors, the VM is deleted,
+the ssh-agent is killed and the UUID of the image is printed. The image is
+ready for publication.
+
+Be aware that if the create script fails, the virtual machine wont be deleted;
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: creation script, will not be

File: scripts/support/create_ge_images/create_template.sh [hunk 8]
```diff
@@ -110,25 +116,25 @@ create_template() {
   # if present, upload file
   if [ -n "${UPLOAD_FILE}" -a -f $IMAGES_DIR/${name}/$UPLOAD_FILE ] ; then
      echo "Uploading ${UPLOAD_FILE}"
-     scp $IMAGES_DIR/${name}/$UPLOAD_FILE ${user}@${IP}:
+     scp $I $IMAGES_DIR/${name}/$UPLOAD_FILE ${user}@${IP}:
   fi
 
   # upload create script
```
Comment (flopezag) [inline scripts/support/create_ge_images/create_template.sh]: creation script

File: scripts/support/create_ge_images/README.rst [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+The local copy of the snapshot is *virt-syspred*. Then the shrink process and
+*virt-sparsify* are invoked. Finally, the resulting image is upload as
+*<imagename>_rc* (*rc* means *release candidate*)
+
+The last phase is to check the image. A server is instantiated, using again the
+generated keypair and assigning the floating IP. This time, however, the security
+group *allopen* is used and the name of the VM is the same than the image but
+with the *test-* prefix. A ssh-agent is started to insert the generated public key.
+Then the test script is invoked. If it complete without errors, the VM is deleted,
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: testing script... If it is completed without errors, the VM with the _test-_ prefix is deleted,

File: scripts/support/create_ge_images/README.rst [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+the ssh-agent is killed and the UUID of the image is printed. The image is
+ready for publication.
+
+Be aware that if the create script fails, the virtual machine wont be deleted;
+this is useful to debug the problem, but anyway that virtual machine should be
+deleted before trying again. In the same way, if the test script fails,
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: testing script

File: scripts/support/create_ge_images/README.rst [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+neither the virtual machine nor the release candidate image are deleted.
+
+**Warning: the test script is executed directly, it is not invoked inside the
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: testing script

File: scripts/support/create_ge_images/README.rst [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+VM machine but in the user account. Therefore, this script is a security risk
+and must be audited before running it. It is important also not using the root
+account to invoke the process. The script clean the sudo credential before
+invoking the script. As an experimental feature, it is possible to run the
+script inside a VM. This is described in next section.
+
+Running the test script in a VM
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: testing script

File: scripts/support/create_ge_images/README.rst [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+-------------------------------
+
+If ``TEST_USING_VM`` is defined and not empty, the test script is executed
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: testing script

File: scripts/support/create_ge_images/create_template.sh [hunk 14]
```diff
@@ -174,31 +180,83 @@ create_template() {
...
+     # Wait until ssh is reading
+     wait_ssh $user
+     # remove sudo credential (to avoid the script use it)
+     sudo -k
+     DIR=$IMAGES_DIR/${name}/nova li
+     eval $(ssh-agent)
+     ssh-add ~/.ssh/createtestimage
+
+     echo "Running the test"
+     # run test script
```
Comment (flopezag) [inline scripts/support/create_ge_images/create_template.sh]: Running the testing script, # run testing script

File: scripts/support/create_ge_images/README.rst [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+.. code::
+
+  /opt/create_ge_images/create_template_debian.sh myimage
+
+Inside the folder myimage, a pair of logs will be generated: *create.log* is the
+result of invoking the *create.sh* script and *test.log* is the output of
+the *test.sh* launch. Of course, if the execution of the create script fails,
+then only *create.log* is generated.
+
+If all the process ends without errors, the UUID of the new image is printed.
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: if all processes end without errors

File: scripts/support/create_ge_images/README.rst [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+At first, the script launch a VM using the chosen base image, with the generated
+keypair and the security group *sshopen*. The name of the virtual machine is
+the same than the image. Then it assigns the floating IP and wait until a SSH
+connection is ready.
+
+At this moment, the create.sh script is uploaded to the server (the data.tgz
+is upload if it exists, too). Then the script is invoked via SSH.
+
+If the script did not fail, the support account and the script are deleted and
+the server is power down. Then a snapshot is made, downloaded and deleted. Also
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: the server is also deleted from the image?

File: scripts/support/create_ge_images/README.rst [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+ready for publication.
+
+Be aware that if the create script fails, the virtual machine wont be deleted;
+this is useful to debug the problem, but anyway that virtual machine should be
+deleted before trying again. In the same way, if the test script fails,
+neither the virtual machine nor the release candidate image are deleted.
+
+**Warning: the test script is executed directly, it is not invoked inside the
+VM machine but in the user account. Therefore, this script is a security risk
+and must be audited before running it. It is important also not using the root
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: not use the root

File: scripts/support/create_ge_images/README.rst [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+account to invoke the process. The script clean the sudo credential before
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: The script cleans

File: scripts/support/create_ge_images/README.rst [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+At this moment, the create.sh script is uploaded to the server (the data.tgz
+is upload if it exists, too). Then the script is invoked via SSH.
+
+If the script did not fail, the support account and the script are deleted and
+the server is power down. Then a snapshot is made, downloaded and deleted. Also
+the VM is deleted.
+
+The local copy of the snapshot is *virt-syspred*. Then the shrink process and
+*virt-sparsify* are invoked. Finally, the resulting image is upload as
+*<imagename>_rc* (*rc* means *release candidate*)
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst:81]: One question, regarding the final name of the image, it includes the version of the GEi?

File: scripts/support/create_ge_images/README.rst [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+Running the test script in a VM
+-------------------------------
+
+If ``TEST_USING_VM`` is defined and not empty, the test script is executed
+inside a VM. This is an experimental feature that does not require an extra
+floating IP. It is more secure, but it needs more time to complete. If the
+script fails, maybe a good idea is to check that this experimental feature is
+not the cause of the problem. To repeat the test only, the variable ``TEST_ONLY``
+must be defined and not empty (and of course to use the traditional method,
+``TEST_USING_VM`` must be undefined)
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: must be undefined).

File: scripts/support/create_ge_images/README.rst [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+
+It works by creating a second VM (the tester). Initially the floating IP is assigned to
+this VM and a SSH connection is created, using SSH ControlMaster; this maintains
+a persistent connection that is reused by the following ssh commands. Then the
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: is reused by the subsequent ssh commands execution.

File: scripts/support/create_ge_images/README.rst [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+IP is assigned to the testing VM; this change does not affect to the already
+created connections. This way is possible to connect to the tester VM via ssh,
+in spite of the floating IP is assigned to the testing VM.
+
+Publishing the image
+********************
+
+Every GE image have two properties: the NID and the type. The values are
+registered at https://wikis.hi.inet/fi-ware/index.php/Cloud:GE_Identification
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst:125]: This is a private url not public, we have to provide a new url in the forje in order to show this data. We have to change it but it should be done during today, together with some improvements. Wait to close the PR.

File: scripts/support/create_ge_images/README.rst (hunk 16 above)
Comment (chemaper) [inline scripts/support/create_ge_images/README.rst]: The server (VM) launched to install the software is deleted after the snapshot, as the test explained:

If the script did not fail, the support account and the script are deleted and
the server is power down. Then a snapshot is made, downloaded and deleted. Also
the VM is deleted.

I'll change server -> VM in the document to make the sentence more clear. Anyway  we use VM more frequently that server in the text (I know that in nova tools the name used is server, but this is not consistent with the own OpenStack documentation where other terms as instance or VM are more frequent)

File: scripts/support/create_ge_images/README.rst (hunk 19 above)
Comment (chemaper) [inline scripts/support/create_ge_images/README.rst:81]: The script uses the image name received as parameter. It is managed as a opaque token (the _rc is added in the creation script but the publish script removes it). 

The image name is provided by the GE owner in the ticket. The same that in the current images, some tickets specify a name with a version an others don't.

File: scripts/support/create_ge_images/README.rst [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+To create an image called *myimage*, the user must create the folder
+*~create_ge_images/myimage* and put inside the scripts *create.sh* and
+*test.sh*. Optionally, the file *data.tgz* can also be added.
+
+It is strongly recommended that both *create.sh* and *test.sh* scripts start with
+``#!/bin/bash -ex``. This way, the script stops when something is wrong and
+a trace is displayed for each command invoked. Anyway, the exit status of the
+script must be zero only when the script runs without error.
+
+After this, one of the scripts *create_template_*.sh*, must be invoked passing
```
Comment (jframos) [inline scripts/support/create_ge_images/README.rst]: You have to escape the '*'.

File: scripts/support/create_ge_images/README.rst [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+
+To publish the image, the script ``publish_image.py`` can be used. The following
+order is used to publish the image *myimage*, that replaces an image already
+existing with the same name:
+
+.. code::
+
+  /opt/create_ge_images/publish_image myimage
+
+The command renames the ol:d image (using the suffix .old) and makes it private.
```
Comment (jframos) [inline scripts/support/create_ge_images/README.rst]: 'old'

File: scripts/support/create_ge_images/create_template.sh [hunk 25]
```diff
@@ -174,31 +180,83 @@ create_template() {
 test_template() {
   name=$1
   user=$2
+  export USERNAME=${user}
+
+  if [ $TEST_USING_VM ] ; then
+     echo "Launching a tester VM"
+     ID2=$(nova boot tester-${name} --security-groups $SECURITY_GROUP_CREATE --flavor m1.tiny --key-name $KEYPAIR --image base_debian_7 --nic net-id=$SHARED_NETWORK_ID | awk -F\| '/\|[ \t]+id[ \t]+\|/ {print $3}')
```
Comment (jframos) [inline scripts/support/create_ge_images/create_template.sh]: more than 120 chars

File: scripts/support/create_ge_images/create_template.sh [hunk 26]
```diff
@@ -174,31 +180,83 @@ create_template() {
...
+     echo "VM id: $ID2"
+  fi
 
   # launch a VM using the new template
   echo "Launching a testing VM"
   ID=$(nova boot test-${name} --security-groups $SECURITY_GROUP_TEST --flavor m1.small --key-name $KEYPAIR --image ${name}_rc --nic net-id=$SHARED_NETWORK_ID | awk -F\| '/\|[ \t]+id[ \t]+\|/ {print $3}')
```
Comment (jframos) [inline scripts/support/create_ge_images/create_template.sh]: more than 120 chars

File: scripts/support/create_ge_images/README.rst [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+  the wiki. The type must be the same that the other images in the same chapter.
+* there is an older version of the image and both will coexist, using for example
+  a suffix with the version number. The NID and type must be provided to the
+  script manually but they must be the same values than the other image.
+* the image replaces an older version, using the same name. The old image
+  will be overwritten. The NID and type of the old image will be used
+  automatically.
+
+The script ``locate_images.py`` can be helpful. It shows a list of the images with
+its name, nid and type. It is also useful to see what NIDs are in use, if a new
```
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst]: I suggest deletion of the content "...if a new one must be used" due to the NID are obtained directly from the node-id in the catalogue.

Comment (chemaper) [inline scripts/support/create_ge_images/README.rst]: Ok. done

File: scripts/support/create_ge_images/README.rst (hunk 19 above)
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst:81]: NTC

File: scripts/support/create_ge_images/README.rst (hunk 22 above)
Comment (flopezag) [inline scripts/support/create_ge_images/README.rst:125]: We leave this issue pending to resolve the next sub-task and then we make a modification in this content.

**Comment by flopezag** [comment]
LGTM

**Pull request merged**